repo/blob/gdrive/gdrive_storage.go

```diff
@@ -0,0 +1,460 @@
+// Package gdrive implements Storage based on Google Drive.
+package gdrive
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	drive "google.golang.org/api/drive/v3"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
+
+	"github.com/kopia/kopia/internal/clock"
+	"github.com/kopia/kopia/internal/iocopy"
+	"github.com/kopia/kopia/internal/retry"
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/blob/retrying"
+)
+
+const (
+	gdriveStorageType = "gdrive"
+	gdriveMimeType    = "application/x-kopia"
+	
+	uploadChunkSize   = 1 << 20
+	uploadContentType = "application/octet-stream"
+	maxPrefixLen      = 26
+	queryRetryDelay   = 100 * time.Millisecond
+	queryRetryMax     = 3
+
+  // googleapi.Field values.
+	listIdFields       = "files(name,id)"
+	listMetadataFields = "files(name,id,mimeType,size,modifiedTime)"
+	getMetadataFields  = "name,id,mimeType,size,modifiedTime"
+)
+
+type gdriveStorage struct {
+	Options
+
+	client      *drive.Service
```

Comment:
I don't know if drive API has something that prevents mocking, but I'm hoping `client` be an interface includes just the methods of `*drive.FilesService` we actually use?

This would allow us to write fake implementation of this interface and unit-test most of this provider without calling actual Google drive, because unlike others it has some pretty non-trivial logic?

I realize this is can be much harder than the current approach (so it perhaps can be done in separate PR, once this early version is merged), but I think it's important to prevent regressions here.

Reply:
Looks like FilesService doesn't mind being stored on its own. Done